File: cxx/src/bridge.cc
// C++ to Rust bridge

#include "bridge.hh"

namespace bridge {
    void spdlog_trace(rust::Str msg) {
        spdlog::trace(std::string_view(msg.data(), msg.size()));
    }

    void spdlog_debug(rust::Str msg) {
        spdlog::debug(std::string_view(msg.data(), msg.size()));
    }

    void spdlog_info(rust::Str msg) {
        spdlog::info(std::string_view(msg.data(), msg.size()));
    }

    void spdlog_warn(rust::Str msg) {
        spdlog::warn(std::string_view(msg.data(), msg.size()));
    }

    void spdlog_error(rust::Str msg) {
        spdlog::error(std::string_view(msg.data(), msg.size()));
    }

    rust::String get_calendar_time() {
        std::array<char, 256> buffer{};
        RE::Calendar::GetSingleton()->GetTimeDateString(buffer.data(), buffer.size(), true);

        return rust::String{ buffer.data() };
    }

    void message_box(rust::Str msg) {
        RE::DebugMessageBox(std::string{ msg }.data());
    }
}  // namespace bridge

File: rust/bridge/src/bridge.rs
#[cxx::bridge(namespace = "bridge")]
pub(crate) mod ffi {
    extern "Rust" {
        fn bridge_init();
    }

    #[allow(unused)]
    unsafe extern "C++" {
        include!("bridge.hh");

        pub fn spdlog_trace(msg: &str);
        pub fn spdlog_debug(msg: &str);
        pub fn spdlog_info(msg: &str);
        pub fn spdlog_warn(msg: &str);
        pub fn spdlog_error(msg: &str);

        pub fn get_calendar_time() -> String;

        /// # Example
        ///
        /// ```rust,no_compile
        /// ffi::message_box("Hello from Rust!");
        /// ```
        pub fn message_box(msg: &str);
    }
}

fn bridge_init() {
    crate::log_trace!("bridge_init() has been called.");
}

#[cfg(test)]
mod tests {
    // For CI
    #[test]
    const fn test_dummy() {
        assert!(2 + 2 == 4);
    }
}
